sympy/codegen/ast.py

```diff
@@ -399,3 +400,405 @@ def body(self):
+      ...
+    ValueError: Casting gives a significantly different value.
+    >>> f64 = Type('float64')
+    >>> f64.cast_check(v10)
+    12345.67894
+    >>> from sympy import Float
+    >>> v18 = Float('0.123456789012345646')
+    >>> f64.cast_check(v18)
+    Traceback (most recent call last):
+      ...
+    ValueError: Casting gives a significantly different value.
+    >>> Type('float80').cast_check(v18)
+    0.123456789012345649
+
+    References
+    ----------
+    [1] https://docs.scipy.org/doc/numpy/user/basics.types.html
+
+    """
+    allowed_names = tuple('intc intp int8 int16 int32 int64 uint8 uint16 uint32'.split() +
+                          'uint64 float16 float32 float64 complex64 complex128'.split() +
+                          'real integer complex bool'.split())
+    __slots__ = ['name']
+
+    default_limits = {  # IEE754 data when applicable
+        'int32': {  # usually the fastest integer typ (and hence 'int' on most systems)
+            'max': 2**31 - 1,  # INT_MAX
+            'min': -2**31  # INT_MIN
+        },
+        'float32': {
+            'max': 3.40282347e+38,  # FLT_MAX
+            'tiny': 1.17549435e-38,  # FLT_MIN, excluding subnormal numbers
+            'eps': 1.1920929e-07,  # FLT_EPSILON
+            'dig': 6,  # FLT_DIG
+            'decimal_dig': 9,  # FLT_DECIMAL_DIG
+        },
+        'float64': {
+            'max': 1.79769313486231571e+308,  # DBL_MAX
+            'tiny': 2.22507385850720138e-308,  # DBL_MIN, excluding subnormal numbers
+            'eps': 2.22044604925031308e-16,  # DBL_EPSILON
+            'dig': 15,  # DBL_DIG
+            'decimal_dig': 17,  # DBL_DECIMAL_DIG
+        },
+        'float80': {  # extended precision, usually "long double", even numpy.float128 (!)
+            'max': 1.18973149535723176502e+4932,  # LDBL_MAX
+            'tiny': 3.36210314311209350626e-4932,  # LDBL_MIN, excluding subnormal numbers
+            'eps': 1.08420217248550443401e-19,  # LDBL_EPSILON
+            'dig': 18,  # LDBL_DIG
+            'decimal_dig': 21,  # LDBL_DECIMAL_DIG
+        }
+    }
+    default_limits['complex64'] = default_limits['float32']
+    default_limits['complex128'] = default_limits['float64']
+
+    default_precision_targets = {  # e.g.:
+        # 'real': 'float64',
+        # 'integer': 'intc',
+        # 'complex': 'complex128'
+    }
+
+    def __new__(cls, name):
+        if isinstance(name, Type):
+            return name
+        return Symbol.__new__(cls, name)
+
+    @classmethod
+    def from_expr(cls, expr, symb=None):
+        """ Infers type from an expression or a ``Symbol``.
+
+        Parameters
+        ----------
+        expr : number, string or SymPy object
+            The typename will be deduced from type or properties. Default is 'real'
+            (e.g. when a string is given as expr).
+        symb : Symbol (optional)
+            If given, assumptions of ``symb`` has higher precedence than expr.
```

Comment:
If I am reading this right, `symb` is not actually used, other than for assumptions. Perhaps we could use the new assumptions here.